pdns/dnsrecords.cc

```diff
@@ -489,6 +489,227 @@ string EUI64RecordContent::getZoneRepresentation(bool noDot) const
 
 /* EUI64 end */
 
+/* APL start */
+/* https://tools.ietf.org/html/rfc3123 */
+void APLRecordContent::report(void)
+{
+  regist(1, QType::APL, &make, &make, "APL");
+}
+
+// Parse incoming packets (e.g. nsupdate)
+std::shared_ptr<DNSRecordContent> APLRecordContent::make(const DNSRecord &dr, PacketReader& pr) {
+  uint8_t temp;
+  APLRDataElement ard;
+  int processed = 0;
+
+  auto ret=std::make_shared<APLRecordContent>();
+
+  while (processed<dr.d_clen) {
+    pr.xfr16BitInt(ard.d_family);
+    pr.xfr8BitInt(ard.d_prefix);
+    pr.xfr8BitInt(temp);
+    ard.d_n = (temp & 128) >> 7;
+    ard.d_afdlength = temp & 127;
+
+    if (ard.d_family == APL_FAMILY_IPV4) { // IPv4
+      if (ard.d_afdlength > 4) {
+        throw MOADNSException("Invalid IP length for IPv4 APL");
+      }
+      bzero(ard.d_ip4, sizeof(ard.d_ip4));
+      // Call min because we can't trust d_afdlength in an inbound packet
+      for (u_int i=0; i < min(ard.d_afdlength, (u_int)4); i++)
+        pr.xfr8BitInt(ard.d_ip4[i]);
+    } else if (ard.d_family == APL_FAMILY_IPV6) {
+      if (ard.d_afdlength > 16) {
+        throw MOADNSException("Invalid IP length for IPv6 APL");
+      }
+      bzero(ard.d_ip6, sizeof(ard.d_ip6));
+      // Call min because we can't trust d_afdlength in an inbound packet
+      for (u_int i=0; i < min(ard.d_afdlength, (u_int)16); i++)
+        pr.xfr8BitInt(ard.d_ip6[i]);
+    } else
+    throw MOADNSException("Unknown family for APL record");
+
+    processed += 4 + ard.d_afdlength;
+
+    ret->aplrdata.push_back(ard);
+  }
+
+  return ret;
+}
+
+// Parse a single APL <apitem>
+APLRDataElement APLRecordContent::parseAPLElement(const string& element) {
+  string record;
+  Netmask nm;
+  int bytes;
+  bool done_trimming;
+  APLRDataElement ard;
+
+  // Parse the optional leading ! (negate)
+  if (element[0] == '!') {
```

Comment:
I know the caller already tested that `element` is not empty but let's prevent a possible issue if that code is ever refactored.
```suggestion
  if (element.at(0) == '!') {
```